### ISSUE TITLE
[picker] does not get field value by slotprops name

### ISSUE THREAD

**Issue by githubtuxedo** [mui/mui-x#8745]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

[Link to live example](https://mui.com/x/api/date-pickers/date-picker/):

```
import { useForm } from "react-hook-form";
const { handleSubmit, register, getValues, setValue } = useForm<FormData>();

    const handleSearch = (params: any) => {
        const formInfo = getValues();
        console.log('result: ', formInfo);
    }

            <form onSubmit={handleSubmit(handleSearch)}>
                <DatePicker
                    slotProps={{
                        textField: {
                            size: 'small',
                            label: 'date',
                            ...register('date'),
                        }
                    }}
                />
                <Button type="submit">test</Button>
            </form>
```

### Current behavior 😯

select date and submit, get result { date: undefined } in handleSearch

### Expected behavior 🤔

select date and submit, get result { date: 'DD/MM/YYYY' }

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by LukasTy** [comment]
Hello. Thank you for creating this issue. 👍 
I've looked into the issue and the root problem seems to be the fact that `react-hook-form` expects the event to be provided to the `onChange` callback, which is of course the case with native input element, but not the case with the `DatePicker` 🤔 
What do you think about using the provided `setValue` function to avoid this issue?
Here is a minimal working example: https://codesandbox.io/s/react-hook-form-example-1ty30g?file=/demo.tsx

**Comment by githubtuxedo** [comment]
> Hello. Thank you for creating this issue. 👍 I've looked into the issue and the root problem seems to be the fact that `react-hook-form` expects the event to be provided to the `onChange` callback, which is of course the case with native input element, but not the case with the `DatePicker` 🤔 What do you think about using the provided `setValue` function to avoid this issue? Here is a minimal working example: https://codesandbox.io/s/react-hook-form-example-1ty30g?file=/demo.tsx

Thank you so much for the solution,  but what if too many picker in a form 🤔 , it's not easy to maintain code

**Comment by LukasTy** [comment]
> Thank you so much for the solution, but what if too many picker in a form 🤔 , it's not easy to maintain code

Indeed, I'd agree that integrating multiple pickers in a form isn't the prettiest thing, but I don't think it's too bad.
Here is an example that I have come up with: https://codesandbox.io/s/react-hook-form-pickers-form-example-jptntj?file=/demo.tsx
What do you think? 🤔 
We'll consider if we should rethink the `onChange` signature to be closer to native events in the future. 😉

**Comment by githubtuxedo** [comment]
> > Thank you so much for the solution, but what if too many picker in a form 🤔 , it's not easy to maintain code
> 
> Indeed, I'd agree that integrating multiple pickers in a form isn't the prettiest thing, but I don't think it's too bad. Here is an example that I have come up with: https://codesandbox.io/s/react-hook-form-pickers-form-example-jptntj?file=/demo.tsx What do you think? 🤔 We'll consider if we should rethink the `onChange` signature to be closer to native events in the future. 😉

sry to bother U again, when i use react-form-hook Controller to control component in mui like picker
```
    const { control } = useForm();
            <Controller
                name="time"
                control={control}
                render={({
                    field: { onChange, onBlur, value, name, ref },
                    fieldState: { invalid, isTouched, isDirty, error },
                    formState,
                  }) => {
                    return (
                        <DatePicker
                            value={value}
                            onChange={(value) => {
                                onChange(value.format('YYYY/MM/DD'))
                            }}
                            // inputRef={ref}
                            
                        />
                    )
                }}
            />
```
when i change the field value , it warn "component is changing the uncontrolled value state of usePickerValue to be controlled.". is there something wrong in base component.
I failed to find a way to solve this, could u do me a favor

**Comment by LukasTy** [comment]
> I failed to find a way to solve this, could u do me a favor

Is your `value` `undefined` initially by any chance? 🤔 
If so, could you try ensuring that the `value` is either something or `null`?

**Comment by githubtuxedo** [comment]
> > I failed to find a way to solve this, could u do me a favor
> 
> Is your `value` `undefined` initially by any chance? 🤔 If so, could you try ensuring that the `value` is either something or `null`?
```
<Controller
                name="time"
                control={control}
                render={({
                    field: { onChange, onBlur, value, name, ref },
                    fieldState: { invalid, isTouched, isDirty, error },
                    formState,
                  }) => {
                    return (
                        <DatePicker
                            value={value??null}
                            onChange={(value) => {
                                onChange(value.format('YYYY/MM/DD'))
                            }}
                        />
                    )
                }}
            />
```
even if i ensure the value is not undefined, it still dosen't work, when i choose a date, get a wrong "te.isBefore is not a function"

**Comment by LukasTy** [comment]
> even if i ensure the value is not undefined, it still dosen't work, when i choose a date, get a wrong "te.isBefore is not a function"

This line causes you problems `onChange(value.format('YYYY/MM/DD'))`. The date pickers starting from v6 no longer support hybrid value binding approaches, where `value` is passed as a string, but internally it still operates with the selected adapter format.
Starting from v6 we expect users to ensure that they are operating with the same date object types `in` and `out` (`value` and `onChange`).
In your case, if you need your dates to be as strings in your preferred format, please consider doing this transformation when you actually need it (probably before submitting a form in your case).

**Comment by githubtuxedo** [comment]
> > even if i ensure the value is not undefined, it still dosen't work, when i choose a date, get a wrong "te.isBefore is not a function"
> 
> This line causes you problems `onChange(value.format('YYYY/MM/DD'))`, simply replacing it with `onChange(value)` avoids your mentioned issue. The date pickers starting from v6 no longer support hybrid value binding approaches, where `value` is passed as a string, but internally it still operates with the selected adapter format. Starting from v6 we expect users to ensure that they are operating with the same date object types `in` and `out` (`value` and `onChange`). In your case, if you need your dates to be as strings in your preferred format, please consider doing this transformation when you actually need it (probably before submitting a form in your case).

Thank you so much,  😉. You are my idol
